File: ui-macos/Sources/ScpCommander/Sites.swift
import Foundation

/// A saved connection — the full session, WinSCP-style: protocol, endpoint,
/// auth method, key file, and S3 bucket/region. Passwords live in the
/// Keychain (only when the user opts in at save time), never in this file.
///
/// Site names may contain "/" to group sites into folders, exactly like
/// WinSCP: "Work/web1" shows as "web1" inside a "Work" folder.
struct Site: Codable, Identifiable, Hashable {
    var id = UUID()
    var name: String
    var proto: Proto
    var host: String
    var port: String
    var user: String
    var authMode: AuthMode
    var keyPath: String
    var bucket: String
    var region: String
    /// Initial directories applied when the site is loaded (WinSCP's
    /// "Remote directory" advanced setting). Empty = defaults.
    var remoteDir: String
    var localDir: String
    /// Optional SFTP bastion (Optional so older saved sites still decode).
    var jumpHost: String?
    var jumpPort: String?
    var jumpUser: String?
    var jumpAuthMode: AuthMode?
    var jumpKeyPath: String?

    init(
        name: String, proto: Proto, host: String, port: String, user: String,
        authMode: AuthMode = .password, keyPath: String = "",
        bucket: String = "", region: String = "",
        remoteDir: String = "", localDir: String = "",
        jumpHost: String? = nil, jumpPort: String? = nil, jumpUser: String? = nil,
        jumpAuthMode: AuthMode? = nil, jumpKeyPath: String? = nil
    ) {
        self.name = name
        self.proto = proto
        self.host = host
        self.port = port
        self.user = user
        self.authMode = authMode
        self.keyPath = keyPath
        self.bucket = bucket
        self.region = region
        self.remoteDir = remoteDir
        self.localDir = localDir
        self.jumpHost = jumpHost
        self.jumpPort = jumpPort
        self.jumpUser = jumpUser
        self.jumpAuthMode = jumpAuthMode
        self.jumpKeyPath = jumpKeyPath
    }

    /// Decode tolerantly so sites.json files from older versions still load.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        name = try c.decode(String.self, forKey: .name)
        proto = try c.decodeIfPresent(Proto.self, forKey: .proto) ?? .sftp
        host = try c.decodeIfPresent(String.self, forKey: .host) ?? ""
        port = try c.decodeIfPresent(String.self, forKey: .port) ?? "22"
        user = try c.decodeIfPresent(String.self, forKey: .user) ?? ""
        authMode = try c.decodeIfPresent(AuthMode.self, forKey: .authMode) ?? .password
        keyPath = try c.decodeIfPresent(String.self, forKey: .keyPath) ?? ""
        bucket = try c.decodeIfPresent(String.self, forKey: .bucket) ?? ""
        region = try c.decodeIfPresent(String.self, forKey: .region) ?? ""
        remoteDir = try c.decodeIfPresent(String.self, forKey: .remoteDir) ?? ""
        localDir = try c.decodeIfPresent(String.self, forKey: .localDir) ?? ""
    }

    /// WinSCP-style folder: the part before the first "/", if any.
    var folder: String? {
        guard let idx = name.firstIndex(of: "/") else { return nil }
        let f = String(name[..<idx])
        return f.isEmpty ? nil : f
    }

    /// Name shown in the list (folder prefix stripped).
    var displayName: String {
        guard let idx = name.firstIndex(of: "/") else { return name }
        return String(name[name.index(after: idx)...])
    }

    var keychainAccount: String {
        Keychain.account(proto: proto, user: user, host: host, port: port)
    }
}

// MARK: - Interchange format (shared with the GTK app)

/// Versioned, human-readable export format. Both the macOS and Ubuntu apps
/// read and write this, so sites can move between machines and platforms.
/// Passwords are intentionally not part of it — they stay in the Keychain.
struct SiteExportFile: Codable {
    var scp_commander_sites = 1
    var sites: [SiteExport]
}

struct SiteExport: Codable {
    var name: String
    var `protocol`: String  // sftp | ftp | ftps | s3
    var host: String
    var port: String
    var user: String
    var auth: String  // password | key | agent
    var key_path: String
    var bucket: String
    var region: String
    var remote_dir: String?
    var local_dir: String?

    init(from site: Site) {
        name = site.name
        `protocol` = ["sftp", "ftp", "ftps", "s3"][Int(site.proto.rawValue) % 4]
        host = site.host
        port = site.port
        user = site.user
        auth = ["password", "key", "agent"][Int(site.authMode.rawValue) % 3]
        key_path = site.keyPath
        bucket = site.bucket
        region = site.region
        remote_dir = site.remoteDir
        local_dir = site.localDir
    }

    var asSite: Site {
        let proto: Proto =
            switch `protocol` {
            case "ftp": .ftp
            case "ftps": .ftps
            case "s3": .s3
            default: .sftp
            }
        let mode: AuthMode =
            switch auth {
            case "key": .keyFile
            case "agent": .agent
            default: .password
            }
        return Site(
            name: name, proto: proto, host: host, port: port, user: user,
            authMode: mode, keyPath: key_path, bucket: bucket, region: region,
            remoteDir: remote_dir ?? "", localDir: local_dir ?? "")
    }
}

/// Saved sites, persisted as JSON under Application Support, kept sorted by
/// name so folder groups stay together.
@MainActor
final class SitesStore: ObservableObject {
    @Published private(set) var sites: [Site] = []

    private let fileURL: URL

    init() {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)
            .first ?? FileManager.default.temporaryDirectory
        let dir = base.appendingPathComponent("ScpCommander", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        fileURL = dir.appendingPathComponent("sites.json")
        load()
    }

    /// Folder names (in display order), with nil for ungrouped sites first.
    var folders: [String?] {
        var seen = [String?]()
        for site in sites {
            if !seen.contains(site.folder) { seen.append(site.folder) }
        }
        // Ungrouped first, then folders alphabetically.
        return seen.sorted { a, b in
            switch (a, b) {
            case (nil, _): return true
            case (_, nil): return false
            case (let x?, let y?): return x.localizedCaseInsensitiveCompare(y) == .orderedAscending
            }
        }
    }

    func sites(in folder: String?) -> [Site] {
        sites.filter { $0.folder == folder }
    }

    func add(_ site: Site) {
        // Replace a same-named entry rather than duplicating.
        if let idx = sites.firstIndex(where: { $0.name == site.name }) {
            sites[idx] = site
        } else {
            sites.append(site)
        }
        sortAndSave()
    }

    func rename(_ site: Site, to newName: String) {
        guard !newName.isEmpty, let idx = sites.firstIndex(where: { $0.id == site.id }) else {
            return
        }
        sites[idx].name = newName
        sortAndSave()
    }

    func remove(_ site: Site) {
        sites.removeAll { $0.id == site.id }
        sortAndSave()
    }

    /// Serialize all sites to the cross-platform interchange format.
    func exportData() throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try encoder.encode(SiteExportFile(sites: sites.map(SiteExport.init)))
    }

    /// Merge sites from interchange data (same-named sites are replaced).
    /// Returns the number of sites in the file.
    func importData(_ data: Data) throws -> Int {
        let file = try JSONDecoder().decode(SiteExportFile.self, from: data)
        for exported in file.sites {
            add(exported.asSite)
        }
        return file.sites.count
    }

    /// Import sessions from a WinSCP.ini file ([Sessions\Name] blocks).
    /// Session names are URL-encoded and may contain "/" folders, which map
    /// straight onto our folder grouping. Returns the number imported.
    func importWinScpIni(_ text: String) throws -> Int {
        struct Pending {
            var name = ""
            var host = ""
            var port: String?
            var user = ""
            var fsProtocol = 0
            var ftpSecure = false
            var keyPath = ""
            var remoteDir = ""
            var localDir = ""
        }

        func decode(_ s: String) -> String {
            s.removingPercentEncoding ?? s
        }

        var current: Pending?

        // Returns 1 when a complete session was flushed into the store.
        // (A local-func + captured-counter version confused the compiler's
        // reachability analysis into a bogus "will never execute" warning.)
        func flush() -> Int {
            guard let p = current, !p.host.isEmpty, !p.name.isEmpty else {
                current = nil
                return 0
            }
            // WinSCP FSProtocol: 5 = FTP (FtpSecure upgrades to FTPS),
            // 7 = S3, everything else is the SSH family → SFTP here.
            let proto: Proto =
                p.fsProtocol == 5 ? (p.ftpSecure ? .ftps : .ftp) : p.fsProtocol == 7 ? .s3 : .sftp
            let port =
                p.port
                ?? {
                    switch proto {
                    case .ftp, .ftps: return "21"
                    case .s3: return "443"
                    default: return "22"
                    }
                }()
            let auth: AuthMode = (!p.keyPath.isEmpty && proto == .sftp) ? .keyFile : .password
            add(
                Site(
                    name: p.name, proto: proto, host: p.host, port: port, user: p.user,
                    authMode: auth, keyPath: p.keyPath,
                    remoteDir: p.remoteDir, localDir: p.localDir))
            current = nil
            return 1
        }

        var count = 0
        for raw in text.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if line.hasPrefix("[") {
                count += flush()
                if line.hasPrefix("[Sessions\\"), line.hasSuffix("]") {
                    let name = String(line.dropFirst("[Sessions\\".count).dropLast())
                    if name != "Default%20Settings" {
                        var pending = Pending()
                        pending.name = decode(name)
                        current = pending
                    }
                }
                continue
            }
            guard current != nil, let eq = line.firstIndex(of: "=") else { continue }
            let key = String(line[..<eq])
            let value = String(line[line.index(after: eq)...])
            switch key {
            case "HostName": current?.host = value
            case "PortNumber": current?.port = value
            case "UserName": current?.user = value
            case "FSProtocol": current?.fsProtocol = Int(value) ?? 0
            case "FtpSecure": current?.ftpSecure = value != "0"
            case "PublicKeyFile": current?.keyPath = decode(value)
            case "RemoteDirectory": current?.remoteDir = decode(value)
            case "LocalDirectory": current?.localDir = decode(value)
            default: break
            }
        }
        count += flush()
        guard count > 0 else {
            throw CoreError(message: "no [Sessions\\…] entries found — is this a WinSCP.ini?")
        }
        return count
    }

    /// Import hosts from an OpenSSH `~/.ssh/config`. Each concrete `Host` alias
    /// (wildcards skipped) becomes an SFTP site grouped under "SSH/", using its
    /// HostName/User/Port/IdentityFile. Returns the number imported.
    func importSshConfig(_ text: String) throws -> Int {
        struct Block {
            var aliases: [String] = []
            var hostName = ""
            var user = ""
            var port = ""
            var identityFile = ""
        }

        // "Key Value" or "Key=Value"; keys are case-insensitive, # comments.
        func parseLine(_ raw: String) -> (key: String, value: String)? {
            var line = raw
            if let hash = line.firstIndex(of: "#") { line = String(line[..<hash]) }
            line = line.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { return nil }
            guard let sep = line.firstIndex(where: { $0 == " " || $0 == "\t" || $0 == "=" })
            else { return (line.lowercased(), "") }
            let key = String(line[..<sep]).lowercased()
            var value = String(line[line.index(after: sep)...])
                .trimmingCharacters(in: CharacterSet(charactersIn: " \t="))
            if value.count >= 2, value.hasPrefix("\""), value.hasSuffix("\"") {
                value = String(value.dropFirst().dropLast())
            }
            return (key, value)
        }

        var blocks: [Block] = []
        var current: Block?
        for raw in text.split(separator: "\n", omittingEmptySubsequences: false) {
            guard let (key, value) = parseLine(String(raw)) else { continue }
            if key == "host" {
                if let c = current { blocks.append(c) }
                var b = Block()
                b.aliases = value.split(whereSeparator: { $0 == " " || $0 == "\t" }).map(String.init)
                current = b
            } else if current != nil {
                switch key {
                case "hostname": current?.hostName = value
                case "user": current?.user = value
                case "port": current?.port = value
                case "identityfile":
                    if current?.identityFile.isEmpty ?? false { current?.identityFile = value }
                default: break
                }
            }
        }
        if let c = current { blocks.append(c) }

        var count = 0
        for b in blocks {
            for alias in b.aliases where !alias.contains("*") && !alias.contains("?") {
                let host = b.hostName.isEmpty ? alias : b.hostName
                guard !host.isEmpty else { continue }
                let key = b.identityFile.isEmpty
                    ? "" : (b.identityFile as NSString).expandingTildeInPath
                add(
                    Site(
                        name: "SSH/\(alias)", proto: .sftp, host: host,
                        port: b.port.isEmpty ? "22" : b.port, user: b.user,
                        authMode: key.isEmpty ? .agent : .keyFile, keyPath: key))
                count += 1
            }
        }
        guard count > 0 else {
            throw CoreError(message: "no Host entries found in the SSH config")
        }
        return count
    }

    private func sortAndSave() {
        sort()
        save()
    }

    private func sort() {
        sites.sort { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    private func load() {
        guard let data = try? Data(contentsOf: fileURL) else { return }
        sites = (try? JSONDecoder().decode([Site].self, from: data)) ?? []
        sort()
    }

    private func save() {
        if let data = try? JSONEncoder().encode(sites) {
            try? data.write(to: fileURL, options: .atomic)
        }
    }
}

File: ui-ubuntu/src/sites.rs
//! Saved connection sites, persisted as JSON under the user config dir
//! (`~/.config/scp-commander/sites.json` on Linux). A site stores the full
//! session, WinSCP-style: protocol, endpoint, auth method, key file, and S3
//! bucket/region. Passwords go to the Secret Service only when the user opts
//! in at save time.
//!
//! Site names may contain "/" to group sites into folders, exactly like
//! WinSCP: "Work/web1" shows as "web1" inside a "Work" folder.

use std::fs;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

#[derive(Clone, Default, Serialize, Deserialize)]
pub struct Site {
    pub name: String,
    /// Index into the protocol dropdown: 0=SFTP, 1=FTP, 2=FTPS, 3=S3.
    pub proto: u32,
    pub host: String,
    pub port: String,
    pub user: String,
    /// Index into the auth dropdown: 0=password, 1=key file, 2=agent.
    #[serde(default)]
    pub auth: u32,
    #[serde(default)]
    pub key_path: String,
    #[serde(default)]
    pub bucket: String,
    #[serde(default)]
    pub region: String,
    /// Initial directories applied when the site is loaded (WinSCP's
    /// "Remote directory" advanced setting). Empty = defaults.
    #[serde(default)]
    pub remote_dir: String,
    #[serde(default)]
    pub local_dir: String,
    /// SFTP bastion (empty jump_host = none). jump_auth: 0 pass, 1 key, 2 agent.
    #[serde(default)]
    pub jump_host: String,
    #[serde(default)]
    pub jump_port: String,
    #[serde(default)]
    pub jump_user: String,
    #[serde(default)]
    pub jump_auth: u32,
    #[serde(default)]
    pub jump_key_path: String,
}

impl Site {
    /// WinSCP-style folder: the part before the first "/", if any.
    pub fn folder(&self) -> Option<&str> {
        match self.name.split_once('/') {
            Some((f, _)) if !f.is_empty() => Some(f),
            _ => None,
        }
    }

    /// Name shown in the list (folder prefix stripped).
    pub fn display_name(&self) -> &str {
        match self.name.split_once('/') {
            Some((f, rest)) if !f.is_empty() => rest,
            _ => &self.name,
        }
    }
}

// --- Interchange format (shared with the macOS app) --------------------------

/// Versioned, human-readable export format. Both the macOS and Ubuntu apps
/// read and write this, so sites can move between machines and platforms.
/// Passwords are intentionally not part of it — they stay in the keyring.
#[derive(Serialize, Deserialize)]
struct SiteExportFile {
    scp_commander_sites: u32,
    sites: Vec<SiteExport>,
}

#[derive(Serialize, Deserialize)]
struct SiteExport {
    name: String,
    protocol: String, // sftp | ftp | ftps | s3
    host: String,
    port: String,
    user: String,
    auth: String, // password | key | agent
    #[serde(default)]
    key_path: String,
    #[serde(default)]
    bucket: String,
    #[serde(default)]
    region: String,
    #[serde(default)]
    remote_dir: String,
    #[serde(default)]
    local_dir: String,
}

impl SiteExport {
    fn from_site(site: &Site) -> Self {
        Self {
            name: site.name.clone(),
            protocol: ["sftp", "ftp", "ftps", "s3"][site.proto as usize % 4].to_string(),
            host: site.host.clone(),
            port: site.port.clone(),
            user: site.user.clone(),
            auth: ["password", "key", "agent"][site.auth as usize % 3].to_string(),
            key_path: site.key_path.clone(),
            bucket: site.bucket.clone(),
            region: site.region.clone(),
            remote_dir: site.remote_dir.clone(),
            local_dir: site.local_dir.clone(),
        }
    }

    fn into_site(self) -> Site {
        let proto = match self.protocol.as_str() {
            "ftp" => 1,
            "ftps" => 2,
            "s3" => 3,
            _ => 0,
        };
        let auth = match self.auth.as_str() {
            "key" => 1,
            "agent" => 2,
            _ => 0,
        };
        Site {
            name: self.name,
            proto,
            host: self.host,
            port: self.port,
            user: self.user,
            auth,
            key_path: self.key_path,
            bucket: self.bucket,
            region: self.region,
            remote_dir: self.remote_dir,
            local_dir: self.local_dir,
            ..Default::default()
        }
    }
}

pub struct SitesStore {
    pub sites: Vec<Site>,
    file: PathBuf,
}

impl SitesStore {
    pub fn load() -> Self {
        let dir = gtk::glib::user_config_dir().join("scp-commander");
        let _ = fs::create_dir_all(&dir);
        let file = dir.join("sites.json");
        let sites = fs::read(&file)
            .ok()
            .and_then(|data| serde_json::from_slice(&data).ok())
            .unwrap_or_default();
        let mut store = Self { sites, file };
        store.sort();
        store
    }

    pub fn add(&mut self, site: Site) {
        // Replace a same-named entry rather than duplicating.
        if let Some(existing) = self.sites.iter_mut().find(|s| s.name == site.name) {
            *existing = site;
        } else {
            self.sites.push(site);
        }
        self.sort();
        self.save();
    }

    pub fn rename(&mut self, index: usize, new_name: &str) {
        if new_name.is_empty() {
            return;
        }
        if let Some(site) = self.sites.get_mut(index) {
            site.name = new_name.to_string();
            self.sort();
            self.save();
        }
    }

    pub fn remove(&mut self, index: usize) {
        if index < self.sites.len() {
            self.sites.remove(index);
            self.save();
        }
    }

    /// Serialize all sites to the cross-platform interchange format.
    pub fn export_interchange(&self) -> Result<String, String> {
        let file = SiteExportFile {
            scp_commander_sites: 1,
            sites: self.sites.iter().map(SiteExport::from_site).collect(),
        };
        serde_json::to_string_pretty(&file).map_err(|e| e.to_string())
    }

    /// Merge sites from interchange data (same-named sites are replaced).
    /// Returns the number of sites in the file.
    pub fn import_interchange(&mut self, data: &str) -> Result<usize, String> {
        let file: SiteExportFile = serde_json::from_str(data).map_err(|e| e.to_string())?;
        let count = file.sites.len();
        for exported in file.sites {
            self.add(exported.into_site());
        }
        Ok(count)
    }

    /// Import sessions from a WinSCP.ini file ([Sessions\Name] blocks).
    /// Session names are URL-encoded and may contain "/" folders, which map
    /// straight onto our folder grouping. Returns the number imported.
    pub fn import_winscp_ini(&mut self, ini: &str) -> Result<usize, String> {
        let mut count = 0;
        let mut current: Option<WinScpSession> = None;
        for raw in ini.lines() {
            let line = raw.trim();
            if line.starts_with('[') {
                if let Some(done) = current.take() {
                    if self.flush_winscp(done) {
                        count += 1;
                    }
                }
                if let Some(name) = line
                    .strip_prefix("[Sessions\\")
                    .and_then(|l| l.strip_suffix(']'))
                {
                    // "Default%20Settings" holds defaults, not a real site.
                    if name != "Default%20Settings" {
                        current = Some(WinScpSession {
                            name: url_decode(name),
                            ..Default::default()
                        });
                    }
                }
                continue;
            }
            let (Some(session), Some((key, value))) = (current.as_mut(), line.split_once('='))
            else {
                continue;
            };
            match key {
                "HostName" => session.host = value.to_string(),
                "PortNumber" => session.port = Some(value.to_string()),
                "UserName" => session.user = value.to_string(),
                "FSProtocol" => session.fs_protocol = value.parse().unwrap_or(0),
                "FtpSecure" => session.ftp_secure = value != "0",
                "PublicKeyFile" => session.key_path = url_decode(value),
                "RemoteDirectory" => session.remote_dir = url_decode(value),
                "LocalDirectory" => session.local_dir = url_decode(value),
                _ => {}
            }
        }
        if let Some(done) = current.take() {
            if self.flush_winscp(done) {
                count += 1;
            }
        }
        if count == 0 {
            return Err("no [Sessions\\…] entries found — is this a WinSCP.ini?".into());
        }
        Ok(count)
    }

    fn flush_winscp(&mut self, s: WinScpSession) -> bool {
        if s.host.is_empty() || s.name.is_empty() {
            return false;
        }
        // WinSCP FSProtocol: 5 = FTP (FtpSecure upgrades to FTPS), 7 = S3,
        // everything else (0/1/2…) is the SSH family → SFTP here.
        let proto = match s.fs_protocol {
            5 => {
                if s.ftp_secure {
                    2
                } else {
                    1
                }
            }
            7 => 3,
            _ => 0,
        };
        let port = s.port.unwrap_or_else(|| {
            match proto {
                1 | 2 => "21",
                3 => "443",
                _ => "22",
            }
            .to_string()
        });
        let auth = if !s.key_path.is_empty() && proto == 0 { 1 } else { 0 };
        self.add(Site {
            name: s.name,
            proto,
            host: s.host,
            port,
            user: s.user,
            auth,
            key_path: s.key_path,
            bucket: String::new(),
            region: String::new(),
            remote_dir: s.remote_dir,
            local_dir: s.local_dir,
            ..Default::default()
        });
        true
    }

    /// Import hosts from an OpenSSH `~/.ssh/config`. Each concrete `Host` alias
    /// (wildcards skipped) becomes an SFTP site grouped under "SSH/", using its
    /// HostName/User/Port/IdentityFile. Returns the number imported.
    pub fn import_ssh_config(&mut self, text: &str) -> Result<usize, String> {
        #[derive(Default)]
        struct Block {
            aliases: Vec<String>,
            host_name: String,
            user: String,
            port: String,
            identity_file: String,
        }
        // "Key Value" or "Key=Value"; keys case-insensitive, # starts a comment.
        fn parse_line(raw: &str) -> Option<(String, String)> {
            let line = raw.split('#').next().unwrap_or("").trim();
            if line.is_empty() {
                return None;
            }
            let sep = line.find([' ', '\t', '='])?;
            let key = line[..sep].to_ascii_lowercase();
            let value = line[sep + 1..]
                .trim_matches([' ', '\t', '='])
                .trim_matches('"');
            Some((key, value.to_string()))
        }

        let mut blocks: Vec<Block> = Vec::new();
        let mut current: Option<Block> = None;
        for raw in text.lines() {
            let Some((key, value)) = parse_line(raw) else { continue };
            if key == "host" {
                if let Some(b) = current.take() {
                    blocks.push(b);
                }
                current = Some(Block {
                    aliases: value.split_whitespace().map(|s| s.to_string()).collect(),
                    ..Default::default()
                });
            } else if let Some(b) = current.as_mut() {
                match key.as_str() {
                    "hostname" => b.host_name = value,
                    "user" => b.user = value,
                    "port" => b.port = value,
                    "identityfile" => {
                        if b.identity_file.is_empty() {
                            b.identity_file = value;
                        }
                    }
                    _ => {}
                }
            }
        }
        if let Some(b) = current.take() {
            blocks.push(b);
        }

        let mut count = 0;
        for b in &blocks {
            for alias in &b.aliases {
                if alias.contains('*') || alias.contains('?') {
                    continue;
                }
                let host = if b.host_name.is_empty() { alias.clone() } else { b.host_name.clone() };
                if host.is_empty() {
                    continue;
                }
                let key_path = expand_tilde(&b.identity_file);
                self.add(Site {
                    name: format!("SSH/{alias}"),
                    proto: 0,
                    host,
                    port: if b.port.is_empty() { "22".into() } else { b.port.clone() },
                    user: b.user.clone(),
                    auth: if key_path.is_empty() { 2 } else { 1 }, // agent | key file
                    key_path,
                    bucket: String::new(),
                    region: String::new(),
                    remote_dir: String::new(),
                    local_dir: String::new(),
                    ..Default::default()
                });
                count += 1;
            }
        }
        if count == 0 {
            return Err("no Host entries found in the SSH config".into());
        }
        Ok(count)
    }

    /// Sorted so folder groups sit together: ungrouped sites first, then
    /// folders alphabetically, each group alphabetical by display name.
    fn sort(&mut self) {
        self.sites.sort_by(|a, b| {
            let key = |s: &Site| {
                (
                    s.folder().is_some(),
                    s.folder().unwrap_or("").to_lowercase(),
                    s.display_name().to_lowercase(),
                )
            };
            key(a).cmp(&key(b))
        });
    }

    fn save(&self) {
        if let Ok(data) = serde_json::to_vec_pretty(&self.sites) {
            let _ = fs::write(&self.file, data);
        }
    }
}

/// Expand a leading `~` in a path against the user's home directory.
fn expand_tilde(path: &str) -> String {
    if path.is_empty() {
        return String::new();
    }
    if let Some(rest) = path.strip_prefix("~/") {
        return gtk::glib::home_dir().join(rest).to_string_lossy().into_owned();
    }
    if path == "~" {
        return gtk::glib::home_dir().to_string_lossy().into_owned();
    }
    path.to_string()
}

/// Accumulator for one [Sessions\…] block while parsing WinSCP.ini.
#[derive(Default)]
struct WinScpSession {
    name: String,
    host: String,
    port: Option<String>,
    user: String,
    fs_protocol: u32,
    ftp_secure: bool,
    key_path: String,
    remote_dir: String,
    local_dir: String,
}

/// Decode WinSCP's %XX escapes (session names, key paths).
///
/// Works purely on bytes: slicing the &str at byte offsets panicked when a
/// `%` was followed by part of a multibyte UTF-8 character (a crafted .ini
/// could crash the app). Embedded NULs are stripped — they'd otherwise break
/// every later CString bridge.
fn url_decode(input: &str) -> String {
    fn hex(b: u8) -> Option<u8> {
        match b {
            b'0'..=b'9' => Some(b - b'0'),
            b'a'..=b'f' => Some(b - b'a' + 10),
            b'A'..=b'F' => Some(b - b'A' + 10),
            _ => None,
        }
    }
    let bytes = input.as_bytes();
    let mut out = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'%' && i + 2 < bytes.len() {
            if let (Some(h), Some(l)) = (hex(bytes[i + 1]), hex(bytes[i + 2])) {
                let value = (h << 4) | l;
                if value != 0 {
                    out.push(value);
                }
                i += 3;
                continue;
            }
        }
        if bytes[i] != 0 {
            out.push(bytes[i]);
        }
        i += 1;
    }
    String::from_utf8_lossy(&out).into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn store() -> SitesStore {
        SitesStore { sites: Vec::new(), file: PathBuf::from("/dev/null") }
    }

    /// Exactly what the macOS app writes — the cross-platform contract.
    #[test]
    fn imports_macos_export() {
        let json = r#"{
            "scp_commander_sites": 1,
            "sites": [
                {
                    "auth": "key",
                    "bucket": "",
                    "host": "example.com",
                    "key_path": "/home/u/.ssh/id_ed25519",
                    "name": "Work/web1",
                    "port": "2222",
                    "protocol": "sftp",
                    "region": "",
                    "user": "deploy"
                }
            ]
        }"#;
        let mut s = store();
        assert_eq!(s.import_interchange(json).unwrap(), 1);
        let site = &s.sites[0];
        assert_eq!(site.proto, 0);
        assert_eq!(site.auth, 1);
        assert_eq!(site.host, "example.com");
        assert_eq!(site.folder(), Some("Work"));
        assert_eq!(site.display_name(), "web1");
    }

    #[test]
    fn imports_winscp_ini() {
        let ini = "\
[Configuration]\nRandomSeedFile=x\n\
[Sessions\\Default%20Settings]\nHostName=ignored\n\
[Sessions\\My%20Server]\nHostName=example.com\nUserName=root\nPortNumber=2222\n\
[Sessions\\Work/web1]\nHostName=web1.example\nUserName=deploy\nFSProtocol=5\nFtpSecure=1\n\
[Sessions\\Keyed]\nHostName=keyed.example\nUserName=ops\nPublicKeyFile=C:%5Ckeys%5Cid.ppk\n";
        let mut s = store();
        assert_eq!(s.import_winscp_ini(ini).unwrap(), 3);
        let by_name = |n: &str| s.sites.iter().find(|x| x.name == n).unwrap();
        let server = by_name("My Server");
        assert_eq!(server.proto, 0);
        assert_eq!(server.port, "2222");
        let web1 = by_name("Work/web1");
        assert_eq!(web1.proto, 2); // FTP + FtpSecure → FTPS
        assert_eq!(web1.port, "21");
        assert_eq!(web1.folder(), Some("Work"));
        let keyed = by_name("Keyed");
        assert_eq!(keyed.auth, 1);
        assert_eq!(keyed.key_path, "C:\\keys\\id.ppk");
    }

    #[test]
    fn imports_ssh_config() {
        let cfg = "\
# my hosts
Host web1 web1.alias
    HostName 10.0.0.5
    User deploy
    Port 2222
    IdentityFile ~/.ssh/id_web

Host *.internal
    User admin

Host db
    HostName db.example.com
";
        let mut s = store();
        // web1 + web1.alias + db = 3; the wildcard block is skipped.
        assert_eq!(s.import_ssh_config(cfg).unwrap(), 3);
        let web1 = s.sites.iter().find(|x| x.name == "SSH/web1").unwrap();
        assert_eq!(web1.host, "10.0.0.5");
        assert_eq!(web1.user, "deploy");
        assert_eq!(web1.port, "2222");
        assert_eq!(web1.auth, 1); // key file
        assert!(web1.key_path.ends_with("/.ssh/id_web"), "got {}", web1.key_path);
        let db = s.sites.iter().find(|x| x.name == "SSH/db").unwrap();
        assert_eq!(db.host, "db.example.com");
        assert_eq!(db.port, "22");
        assert_eq!(db.auth, 2); // agent (no IdentityFile)
    }

    #[test]
    fn url_decode_is_panic_proof() {
        // Regression: "%aé" used to panic (str slice inside a multibyte char).
        assert_eq!(url_decode("%a\u{e9}"), "%a\u{e9}");
        assert_eq!(url_decode("My%20Site"), "My Site");
        assert_eq!(url_decode("a%00b"), "ab"); // NULs stripped
        assert_eq!(url_decode("%"), "%");
        assert_eq!(url_decode("%zz"), "%zz");
    }

    #[test]
    fn export_import_round_trip() {
        let mut s = store();
        s.sites.push(Site {
            name: "S3/backups".into(),
            proto: 3,
            host: "minio.local".into(),
            port: "443".into(),
            user: "AKIA123".into(),
            auth: 0,
            key_path: String::new(),
            bucket: "backups".into(),
            region: "us-east-1".into(),
            remote_dir: String::new(),
            local_dir: String::new(),
            ..Default::default()
        });
        let json = s.export_interchange().unwrap();
        let mut other = store();
        assert_eq!(other.import_interchange(&json).unwrap(), 1);
        assert_eq!(other.sites[0].proto, 3);
        assert_eq!(other.sites[0].bucket, "backups");
        // Importing again replaces rather than duplicating.
        assert_eq!(other.import_interchange(&json).unwrap(), 1);
        assert_eq!(other.sites.len(), 1);
    }
}
